packages/server/modules/workspaces/services/workspaceProjects.ts

```diff
@@ -0,0 +1,31 @@
+import { StreamRecord } from '@/modules/core/helpers/types'
+import { getUserStreams as repoGetUserStreams } from '@/modules/core/repositories/streams'
+import { GetAllWorkspaceProjectsForUser } from '@/modules/workspaces/domain/operations'
+
+/**
+ * TODO: Because of the way we implicitly grant project roles for users in a workspace,
+ * this is equivalent to "get all workspace projects." We may need a new repo function if we ever
+ * need to get workspace projects in a context where we don't have a user id available.
+ */
+export const getAllWorkspaceProjectsForUserFactory =
+  ({
+    getUserStreams
+  }: {
+    getUserStreams: typeof repoGetUserStreams
+  }): GetAllWorkspaceProjectsForUser =>
+  async ({ userId, workspaceId }) => {
+    let cursor: string | undefined = undefined
+    const projects: StreamRecord[] = []
+
+    do {
+      const { streams: workspaceProjects, cursor: pageCursor } = await getUserStreams({
+        userId,
+        workspaceIdWhitelist: [workspaceId],
+        cursor
+      })
+      projects.push(...workspaceProjects)
+      cursor = pageCursor ?? undefined
+    } while (!!cursor)
```

Comment:
1. while/do-while is generally scary and should always have some kind of failsafe, e.g. always breaking after X loops. X can be something practically unreachable like 500 pages of projects, but at the very least it ends somewhere instead of executing indefinitely
2. just infinitely batching up all projects as long as there are more is likely to be a performance bomb - there could be a shit ton of projects OR there could be a bug in the repo query causing an infinitely increasing array of projects. instead, use JS generators and yield each page individually so that we never have to keep more than 1 page in memory - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator